Update the API according to discussion in #82 and #78

zolkis opened the pull request:
Don't merge yet, this is work in progress.
For the rendered version of the tip of the PR, use [this link](http://rawgit.com/zolkis/wot-scripting-api/master/index.html).

Signed-off-by: Zoltan Kis <zoltan.kis@intel.com>

mkovatsc commented:
This PR introduces many completely new concepts that conflict with our long-lived discussion in #82, e.g.:

* expose vs produce
* typedef ThingModel and reinterpretation of ThingTemplate
* ExposedThing without ConsumedThing interface (no control over the Thing)
* Cut down Interaction Inits
* dictionary ValueDefinition
* Additional set*Handler() functions

Is this to be changed as part of the work-in-progress, or are you proposing this now?

zolkis commented:
We have addressed these points during the call. Now the proposal is more or less clean. The reasons for change are the following.

1. expose vs produce
This is mainly a name change, coming from the fact that the returned object (ExposedThing) is not yet an exposed Thing, but an object that needs to  or might undergo changes such as adding properties/actions/events and request handlers. Please read the text, the current proposal allows creating an ExposedThing in 3 ways:
- based on a model (name + semantic annotations) that needs to be modified by adding properties/actions/events/handlers
- based on a TD, that needs to be added handlers
- based on a ConsumedThing (actually its TD, but implementations may do optimizations), that needs to be added handlers.

2. ThingModel reflects the possibilities from above, including ThingTemplate (that remained the same and in #82).

3. Actually a script has total control over an ExposedThing. It defines the thing, so cannot use it before it is defined. If ConsumedThing interface is needed, it needs to consume the ExposedThing's generated TD.

4.
> Cut down Interaction Inits

I am not sure what do you mean here. Perhaps 6. explains.

5. ValueDefinition is an attempt to align with the TD spec. Before it a full ThingDescription was used for specifying types - which was wrong and confusing.

6. The set*Handler() functions are needed for two reasons:
- it is an antipattern to specify function in init dictionary members
- they make explicit the decision that a TD does not contain scripts. Before, if you wanted to create and ExposedThing based on a template/TD, and wanted to add code (that was not present in the template/TD) you first need to remove e.g. a property/action, then re-add it with the handler mentioned in the init dict. This was just wrong. Now you can take a template/TD/ConsumedThing, replicate the TD-related structures and add the handlers on top. Also, it is clear at the top level that an ExposedThing must specify handlers, it is not any more hidden behind the members of a parameter of a function.

In fact, part of these changes are small, part of them are needed. The question is consistency, usability and alignment with the TD spec and other APIs. I believe this version is better (if not perfect) in that sense.

mkovatsc commented:
Mainly, we must not repeat the mishap of a broken (FP)WD, which was caused by ad hoc changes and introduction of unfinished, unconsolidated concepts. So why introduce these changes now and independently? The conclusion in the call was to write up what was discussed there.

1. expose vs produce -> fine by me
  Okay, small rename to be more precise, but also potentially more confusing in regard to established terminology.
2. ThingModel -> partly fine
  Ideally, ThingTemplate and ThingDescription would become the same; also only a subset of a TD is valid for exposing Things (e.g., not the binding information). Thus, `ThingTemplate` was introduced as intermediary solution. For now, we could allow for an object (ThingTemplate) or a string (ThingDescription), which is similar to an earlier version of the API.
  To put a ConsumedThing there seems to only fulfill a specific use case of optimizing proxy Thing creation and opens the path for the strange pattern of 3. This should not be introduced, definitely not for Prague because of the implications on the necessary implementation.
3. ExposedThing control -> not fine by me
  It is pretty ugly to juggle two objects, an ExposedThing and a ConsumedThing, to realize a fundamental use case: implementing the internals of a Thing. The `writeProperty()` function of an ExposedThing is different from the one of ConsumedThing: an ExposedThing is under the local control of the code and directly manipulates the state, while a ConsumedThing only sends request to modify a remote Thing under control of someone else. It was convenient to let ExposedThing implement the interface of ConsumedThing, but this example here tells me, it would be clearer to fully specify both interfaces independently. This was also raised in one of the calls with a similar motiviation.
4. Interaction Inits -> might be fine
  Weird that this comes up only now, as we are basically going back to the initial API (cf. `onInvokeAction()` etc.). This in itself might be a good sign, though.

mkovatsc commented:
@knimura and @danielpeintner please comment, so we can have a prepared call on Monday.

zolkis commented:
> 1. expose vs produce -> fine by me 

The remaining question is whether should we rather make it a constructor to ExposedThing instead of a factory method with disputed naming? Then we avoid the terminology discussion and have cleaner, more JS-like definition to create local ExposedThing objects. The only reason to keep the factory method would be if we wanted to create remote Things, but that is no longer a use case, as it can be done as a service.

> 2. ThingModel -> partly fine

I am also fine with removing ConsumedThing from the models. I was just thinking it could offer significant optimization opportunity when we want to re-expose a consumed thing, i.e. we don't need to fetch/consume a TD, then read the serialized form from the consumed thing and feed it to the ExposedThing factory that will parse it again. However, we could add this later, eventually.

> 3. ExposedThing control -> not fine by me

I am also fine with ExposedThing extending ConsumedThing, but we should be clear about the use cases.
The ExposedThing will define the properties, actions and events, then how to update/read a property, how to call an action, when to emit an event etc. So why would it want to use ConsumedThing methods to call an action, update/read a property and attach event listeners? It is implementing and exposing them, not using them as a client, not until finished exposing. There are two stages here, 1. exposing, 2. using (after consuming). I thought to make that distinction more clear. Perhaps we need to think what is the best way to expose the ConsumedThing interface on an ExposedThing, and when.

> 4. Interaction Inits -> might be fine

After playing with examples, I am pretty convinced now that this is the best way to go. I should add more example code to the spec.

mkovatsc commented:
1. Postpone discussing the constructor idea after Prague or prepare options for a F2F breakout.
2. Good, then let's limit to ThingTemplate dictionary or TD string. ConsumedThing.description can be passed and does not need a fetch.
3. It's not about extending ConsumedThing, but implementing the interface, as we need similar functions to control the local state. Consuming usually happens in a remote runtume. The implementation of the functions will be fundamentally different, yet I don't think we will need different names. At least postpone this discussion.
4. Let's bounce this of the other editors before the call.

knimura commented:
> 1. expose vs produce

No issues with the name change.

> 2. ThingModel – what is ThingModel

Not clear the definition of ThingModel. I have no problem with ThingTemplate, if the definition of ThingTemplate is the complete subset of TD but just excluding binding information from it.

> 3. “Actually a script has total control over an ExposedThing. It defines the thing, so cannot use it before it is defined. If ConsumedThing interface is needed, it needs to consume the ExposedThing's generated TD.”

I agree with ExposedThing extending ConsumedThing, because ExposedThing should have the same function as ConsumedThing. But maybe difficult to understand the notion of "inheriting client API of local thing is part of ExposedThing(ServerAPI)", so showing the use case would be needed.

> 5. ValueDefinition is an attempt to align with the TD spec. Before this a full ThingDescription was used for specifying types - which was wrong and confusing.”

Totally agree.


> 6.	“The set*Handler() functions are needed for two reasons:
> If you wanted to create and ExposedThing based on a template/TD, and wanted to add code you first needed to remove e.g. a property/action, then re-add it with the handler mentioned in the init dict. 
	
Agree. If ThingTemplate is subset of TD and ExposedThing is created from ThingTemplate, both removeProperty and addProperty must be called to setting handler. If we feel that going back to the initial API, introducing UpdateProperty may be one of the solution to avoid onRetrieveProperty or so.

zolkis commented:
> Not clear the definition of ThingModel. 

A Thing model (the term was taken from #64) is either a ThingTemplate (dictionary) or a ThingDescription (string).

I have changed back to `ExposedThing implements ConsumedThing`.

zolkis commented:
As discussed on the call, on request by @knimura , added a short informal section for Observables, and on  request by @danielpeintner , removed optionality from some of the arguments and dictionary members.

zolkis commented:
As discussed in the call, added explanations on how handlers are working/implemented on ExposedThing. @mjkoster please take a look.

danielpeintner commented:
One comment/question: I noticed that PropertyWriteHandler  gets only the new value. We used to have "oldValue" and "newValue". Any rational for this change?

zolkis commented:
WiP is removed, proof-reading can start.

mkovatsc commented:
> One comment/question: I noticed that PropertyWriteHandler gets only the new value. We used to have "oldValue" and "newValue". Any rational for this change?

This is a quite strong concept change. As I mentioned, we had a reactive programming pattern before, which has at least two benefits:

* Script developers do not need to care about proper and safe state management. Without setting a PropertyHandlers, we have a basic resource behavior (set state with PUT, read state with GET).
* Better performance, as requests can be answered directly from the runtime space, no need to hand over to the scripting space (which was a bit problematic with our previous Java + Nashorn implementation and will be problematic for embedded runtimes).

Such a change must be explicitly discussed and is not fit for Prague.

zolkis commented:
There is an editor's Note in the spec stating:

> Note that this function is invoked by implementations before the property is updated, so the client code in this function can invoke the readProperty() method to find out the old value of the property, if needed. Therefore the old value is not provided to this method.

So your use case seems to be solved.

mkovatsc commented:
We are discussing server side -- the client has nothing to do with these handlers.

zolkis commented:
Yes, server side. Client there means client to the server implementation (servient), i.e. the "guest code" in the callback.

mkovatsc commented:
Okay, I see you already removed "client" from that note ;)

Have you checked that the handler code has always the right reference to the ExposedThing object to call `readProperty(propertyName)`? To me it is not obvious without testing, also function vs lambda...

zolkis commented:
Of course the handler code has the right reference to the ExposedThing object.
```javascript
thing.setPropertyWriteHandler(value => {
  thing.readProperty(propertyName).then(oldValue => {
    // put handler code here
   if (oldValue < 5) {
      // update to value 
      gpio.port.write(0x02, value);
   } 
  });
}, propertyName);
```

danielpeintner commented on index.html:
```diff
@@ -262,29 +240,60 @@ <h2>The <dfn>WoT</dfn> object</h2>
             USVString url;
-            Dictionary description;
+            USVString query;
+            sequence&lt;Dictionary&gt; constraints;
```
I wonder why we need  a `sequence<Dictionary>`.
According to my understanding `Dictionary constraints` seems sufficient.

danielpeintner commented on index.html:
```diff
@@ -262,29 +240,60 @@ <h2>The <dfn>WoT</dfn> object</h2>
             USVString url;
-            Dictionary description;
+            USVString query;
+            sequence&lt;Dictionary&gt; constraints;
```
The AND and OR relationship for constraints seems to be very powerful but also complex at first. Moreover, the link to "the semantic metadata" in its description is very fuzzy. As I understand it, the key of the key-value pair is a reference to `SemanticType`. If so we might want to add an example.

zolkis replied:
Example 4 is such an example: discover nearby things by Bluetooth OR NFC.

danielpeintner commented on index.html:
```diff
@@ -293,80 +302,160 @@ <h4>The <dfn>ThingFilter</dfn> dictionary</h4>
+            from a <a>Thing Description</a>, then adding request handlers;
+          </li>
+          <li>
+            from an existing <a>ConsumedThing</a> object, based on its <a>Thing Description</a>, then adding request handlers.
```
The third option is a special case of option 2. Hence I propose to remove option 3 and change option 2 to 

from a Thing Description (possibly also retrieved from a ConsumedThing), then adding request handlers;

zolkis replied:
That should be fine. I just wanted to highlight it is possible to expose a ConsumedThing.

zolkis replied:
Done (not committed yet).

danielpeintner commented on index.html:
```diff
@@ -817,32 +723,26 @@ <h4>The <dfn>SemanticType</dfn> dictionary</h4>
-            Function action;
+          dictionary ThingActionInit: SemanticAnnotations {
+            required DOMString name;
+            ValueType inputDataDescription;
```
Align with TD draft, name it `inputData` and `outputData`?

zolkis replied:
`inputData` would imply this is data, but it is a description of input data.

danielpeintner commented on index.html:
```diff
@@ -853,105 +753,80 @@ <h4>The <dfn>ThingActionInit</dfn> dictionary</h4>
-            ThingDescription dataDescription;
+          dictionary ThingEventInit: SemanticAnnotations {
+            required DOMString name;
+            ValueType dataDescription;
```
Align with TD draft, name it `outputData`?

zolkis replied:
`outputData` is a bad name, usually events have "data" in JS. This API is for developers.

danielpeintner commented on index.html:
```diff
@@ -962,31 +837,46 @@ <h2>Examples</h2>
       <pre class="example highlight" title="Create a new blank exposed Thing">
         WoT.createExposedThing(thingDescription)
```
Old API example. Should be `WoT.produce` without Promise and `then` etc

Also in addAction there is still old handler code.

zolkis replied:
Thanks for the catch. Done.

zolkis commented on index.html:
```diff
@@ -293,80 +302,160 @@ <h4>The <dfn>ThingFilter</dfn> dictionary</h4>
+            from a <a>Thing Description</a>, then adding request handlers;
+          </li>
+          <li>
+            from an existing <a>ConsumedThing</a> object, based on its <a>Thing Description</a>, then adding request handlers.
```
Done.

zolkis commented on index.html:
```diff
@@ -293,80 +302,160 @@ <h4>The <dfn>ThingFilter</dfn> dictionary</h4>
+          url: "http://directory.wotservice.org"
+        };
         let subscription = wot.discover(discoveryFilter).subscribe(
           thing => { console.log("Found Thing " + thing.url); },
```
Done.

zolkis commented on index.html:
```diff
@@ -262,29 +240,60 @@ <h2>The <dfn>WoT</dfn> object</h2>
             USVString url;
-            Dictionary description;
+            USVString query;
+            sequence&lt;Dictionary&gt; constraints;
```
See Example 4. ("protocol" is made up for the example)

zolkis commented on index.html:
```diff
@@ -375,15 +464,28 @@ <h4>The <dfn>ThingInit</dfn> dictionary</h4>>
+      <pre class="example" title="Discover Things exposed nearby, via Bluetooth or NFC">
+        let subscription = wot.discover({
+          method: "nearby",
+          constraints: [{ protocol: "BLE-4.2" }, { protocol: "NFC"}]
```
These examples are made up, at the moment we don't have a standard attribute "protocol", but it could be done if it made sense in the Thing context.

zolkis replied:
Yes,  the constraints MAY be of type `SemanticAnnotation` (type + value), but in general they don't need to be. They are (at the moment) key -value pairs (dictionaries) that can be mapped to `SemanticAnnotation` like 
```javascript
let constraint = { 
    semanticTypes: [ { name: "protocol", context: "..." } ], 
    metadata: [ { type: "protocol", value: "BLE-4.2" } ]
};
```
which is very developer-unfriendly way to tell a constraint. Instead, the implementation of the Scripting API can translate the constraint dictionaries to SemanticAnnotation.

zolkis commented on index.html:
```diff
@@ -395,30 +497,26 @@ <h4>The <dfn>ThingInit</dfn> dictionary</h4>>
         readonly attribute DOMString name;
-        readonly attribute USVString url;
-        readonly attribute ThingDescription description;
+        readonly attribute ThingDescription td;
```
A property has a getter function that can be the `getTd()` - in my view they are equivalent, aren't they?
Of course we could even have a `Promise<ThingDescription> getTD()`, but do we need it?

zolkis commented on index.html:
```diff
@@ -634,111 +599,76 @@ <h2>Examples</h2>
       };
-      ExposedThing implements ConsumedThing;
+      callback ActionHandler = Promise&lt;any&gt; (any parameters);
+      callback PropertyReadHandler = Promise&lt;any&gt; (DOMString propertyName);
```
Right, we discussed this and I haven't removed them. Done now.

zolkis commented on index.html:
```diff
@@ -749,20 +679,22 @@ <h3>The <dfn>RequestHandler</dfn> callback</h3>
+            required DOMString name;
+            required ValueType type;
+            any value;
             boolean configurable = false;
```
Done.

zolkis commented:
There is still a lot of work needed on SemanticAnnotations and DataSchema, but for the moment this is good to be merged.

zolkis commented:
I get a ReSpec warning for undefined Observable, but it's defined. Someone please take another look and try to fix.

danielpeintner commented:
Merging is fine by me also!

Note: the ReSpec issue you are seeing relates to the observable attribute in "5.7.1 The ThingPropertyInit dictionary". The name collision seems to cause that issue. Renaming the attribute made the ReSpec issue disappear.

Let's try to look into after merging...

zolkis commented:
This is merged now. Thank you all for the comments and reviews.
